File: leetCode/medium/1637-widestVerticalAreaBetweenTwoPoints.ts
import { assertEquals } from "https://deno.land/std@0.182.0/testing/asserts.ts";
import {
  describe,
  it,
  beforeEach,
  afterEach,
} from "https://deno.land/std@0.182.0/testing/bdd.ts";

function widestVerticalAreaBetweenTwoPoints(points: Array<[number, number]>) {
  const xCoordsSorted = points.map((point) => point[0]).sort((a, z) => a - z);

  let left = 0;
  let right = 1;
  const coordsLen = xCoordsSorted.length;
  const distanceArr: number[] = [];

  while (right < coordsLen && left < right) {
    const leftXCoord = xCoordsSorted[left];
    const rightXCoord = xCoordsSorted[right];
    distanceArr.push(rightXCoord - leftXCoord);

    left += 1;
    right += 1;
  }

  return Math.max(...distanceArr);
}

export { widestVerticalAreaBetweenTwoPoints };

describe("widestVerticalAreaBetweenTwoPoints", () => {
  it("1. should return the widest vertical area between two points", () => {
    const points: Array<[number, number]> = [
      [8, 7],
      [9, 9],
      [7, 4],
      [9, 7],
    ];
    assertEquals(widestVerticalAreaBetweenTwoPoints(points), 1);
  });

  it("2. should return the widest vertical area between two points", () => {
    const points: Array<[number, number]> = [
      [3, 1],
      [9, 0],
      [1, 0],
      [1, 4],
      [5, 3],
      [8, 8],
    ];
    assertEquals(widestVerticalAreaBetweenTwoPoints(points), 3);
  });
});

File: leetCode/medium/1637-widestVerticalAreaBetweenTwoPoints.jl
function widestVerticalAreaBetweenTwoPoints(points)
  xCoordsSorted = map(point -> point[1], points) |> sort

  left = 1
  right = 2
  coordsLen = length(xCoordsSorted)
  distanceArr = []

  while right <= coordsLen && left < right
    leftXCoord = xCoordsSorted[left]
    rightXCoord = xCoordsSorted[right]
    push!(distanceArr, rightXCoord - leftXCoord)

    left += 1
    right += 1
  end

  return maximum(distanceArr)
end

export widestVerticalAreaBetweenTwoPoints

using Test

Test.@testset "Example 1" begin
  points = [[8, 7], [9, 9], [7, 4], [9, 7]]
  Test.@test widestVerticalAreaBetweenTwoPoints(points) == 1

  points = [[3, 1], [9, 0], [1, 0], [1, 4], [5, 3], [8, 8]]
  Test.@test widestVerticalAreaBetweenTwoPoints(points) == 3
end